In [1]:
# Task 1 — Fix Hidden Missing Values
import numpy as np
import pandas as pd

df = pd.read_csv('adult.csv')
df.columns = df.columns.str.replace('.', '-', regex=False)
df.info()
print(df.head())
print('Null counts before replacement:')
print(df.isnull().sum())

# Replace the missing-value marker from the task.
df.replace(' ?', np.nan, inplace=True)
# This CSV uses '?' without a leading space, so replace that form too.
df.replace('?', np.nan, inplace=True)
print('Null counts after replacement:')
print(df.isnull().sum())

<class 'pandas.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             32561 non-null  int64
 1   workclass       32561 non-null  str  
 2   fnlwgt          32561 non-null  int64
 3   education       32561 non-null  str  
 4   education-num   32561 non-null  int64
 5   marital-status  32561 non-null  str  
 6   occupation      32561 non-null  str  
 7   relationship    32561 non-null  str  
 8   race            32561 non-null  str  
 9   sex             32561 non-null  str  
 10  capital-gain    32561 non-null  int64
 11  capital-loss    32561 non-null  int64
 12  hours-per-week  32561 non-null  int64
 13  native-country  32561 non-null  str  
 14  income          32561 non-null  str  
dtypes: int64(6), str(9)
memory usage: 3.7 MB
   age workclass  fnlwgt     education  education-num marital-status  \
0   90         ?   77053       HS-grad              9

In [2]:
# Task 2 — Quantify Missingness
missing_report = pd.DataFrame({
    'missing_count': df.isnull().sum(),
    'missing_percentage': df.isnull().mean().mul(100)
}).sort_values('missing_count', ascending=False)
print(missing_report)
print('The most affected columns are occupation, workclass, and native-country.')

                missing_count  missing_percentage
occupation               1843            5.660146
workclass                1836            5.638647
native-country            583            1.790486
fnlwgt                      0            0.000000
education                   0            0.000000
education-num               0            0.000000
age                         0            0.000000
marital-status              0            0.000000
relationship                0            0.000000
sex                         0            0.000000
race                        0            0.000000
capital-gain                0            0.000000
capital-loss                0            0.000000
hours-per-week              0            0.000000
income                      0            0.000000
The most affected columns are occupation, workclass, and native-country.


In [3]:
# Task 3 — Handle Missing Values
# Fill missing categorical values with 'Unknown' so no rows are lost.
for column in ['workclass', 'occupation', 'native-country']:
    df[column] = df[column].fillna('Unknown')

print('Remaining missing values:', int(df.isnull().sum().sum()))
print('There are no missing numeric values, so numeric filling is not needed.')

Remaining missing values: 0
There are no missing numeric values, so numeric filling is not needed.


In [4]:
# Task 4 — Detect Duplicates
exact_duplicates = df.duplicated().sum()
feature_columns = df.columns.drop('income')
duplicates_without_target = df.duplicated(subset=feature_columns).sum()
print('Exact duplicate rows:', exact_duplicates)
print('Duplicates ignoring income:', duplicates_without_target)
print('Decision: only exact duplicates count as duplicate records. Rows with identical features')
print('but different income labels are conflicting observations and should be investigated, not dropped automatically.')

Exact duplicate rows: 24
Duplicates ignoring income: 25
Decision: only exact duplicates count as duplicate records. Rows with identical features
but different income labels are conflicting observations and should be investigated, not dropped automatically.


In [5]:
# Task 5 — Fix Inconsistent Categorical Data
columns_to_check = ['education', 'marital-status', 'native-country']

# Remove leading and trailing spaces first.
for column in columns_to_check:
    df[column] = df[column].str.strip()

# Now check the unique values.
for column in columns_to_check:
    print(f'\nUnique values in {column}:')
    print(df[column].unique())


Unique values in education:


<StringArray>
[     'HS-grad', 'Some-college',      '7th-8th',         '10th',
    'Doctorate',  'Prof-school',    'Bachelors',      'Masters',
         '11th',   'Assoc-acdm',    'Assoc-voc',      '1st-4th',
      '5th-6th',         '12th',          '9th',    'Preschool']
Length: 16, dtype: str

Unique values in marital-status:
<StringArray>
[              'Widowed',              'Divorced',             'Separated',
         'Never-married',    'Married-civ-spouse', 'Married-spouse-absent',
     'Married-AF-spouse']
Length: 7, dtype: str

Unique values in native-country:
<StringArray>
[             'United-States',                    'Unknown',
                     'Mexico',                     'Greece',
                    'Vietnam',                      'China',
                     'Taiwan',                      'India',
                'Philippines',            'Trinadad&Tobago',
                     'Canada',                      'South',
         'Holand-Netherlands',           

In [6]:
# Task 6 — Descriptive Statistics
print(df.describe())
print(df.describe(include=['object', 'string']))
summary_stats = df[['age', 'hours-per-week']].agg(['mean', 'median'])
print(summary_stats)
print('Age: mean 38.58 vs median 37.00 (a small difference, suggesting mild right skew).')
print('Hours/week: mean 40.44 vs median 40.00 (not meaningfully different).')

                age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  
       workclass education      marital-status      occupation 

In [7]:
# Task 7 — Frequency Analysis
print('Most common occupation:')
print(df['occupation'].value_counts().head(1))

print('\nSex distribution (%):')
print(df['sex'].value_counts(normalize=True) * 100)

print('\nIncome distribution (%):')
print(df['income'].value_counts(normalize=True) * 100)

print('The income classes are imbalanced: about 76% are <=50K and 24% are >50K.')

Most common occupation:
occupation
Prof-specialty    4140
Name: count, dtype: int64

Sex distribution (%):
sex
Male      66.920549
Female    33.079451
Name: proportion, dtype: float64

Income distribution (%):
income
<=50K    75.919044
>50K     24.080956
Name: proportion, dtype: float64
The income classes are imbalanced: about 76% are <=50K and 24% are >50K.


In [8]:
# Task 8 — Cross-Check a Data Quality Assumption
education_mapping = df.groupby('education')['education-num'].unique()
print(education_mapping)

issues = 0
for education, numbers in education_mapping.items():
    if len(numbers) > 1:
        print('Inconsistency:', education, numbers)
        issues = issues + 1

print('Number of inconsistencies:', issues)

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education-num, dtype: object


Number of inconsistencies: 0


In [9]:
# Task 9 — MySQL-Based Data Extraction
from sqlalchemy import create_engine, text

engine = create_engine('mysql+pymysql://adult_lab:AdultLab_2026!@localhost:3306/adult_income_db')
df.to_sql('adult_income', engine, if_exists='replace', index=False)
sql_all_df = pd.read_sql_query(text('SELECT * FROM adult_income'), engine)
over_30_df = pd.read_sql_query(text('SELECT * FROM adult_income WHERE age > 30'), engine)
engine.dispose()

print('DataFrame shape:', df.shape)
print('MySQL full-table shape:', sql_all_df.shape)
print('MySQL age > 30 shape:', over_30_df.shape)
print('Full extraction correct:', sql_all_df.shape == df.shape)
print(over_30_df.head())

DataFrame shape: (32561, 15)
MySQL full-table shape: (32561, 15)
MySQL age > 30 shape: (21989, 15)
Full extraction correct: True
   age workclass  fnlwgt     education  education-num marital-status  \
0   90   Unknown   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66   Unknown  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital-gain  \
0            Unknown  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2            Unknown      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital-loss  hours-per-we

# Task 10 — Data Profiling Summary Report

**Dataset overview:** The Adult Income dataset contains **32,561 rows and 15 columns**. It describes demographic, education, employment, and financial attributes used to classify whether annual income is `<=50K` or `>50K`.

**Data quality issues:** Hidden text markers represented 4,262 missing values: occupation had 1,843 (5.66%), workclass 1,836 (5.64%), and native-country 583 (1.79%). There were 24 exact duplicate rows and 25 duplicates when income was ignored, meaning one additional pair shared attributes but had different labels. Whitespace normalization was applied to categorical text.

**Key observations:** Mean/median age were 38.58/37 years, while mean/median work time were 40.44/40 hours per week. Prof-specialty was the most common known occupation (4,140 rows). The sample was 66.92% male and 33.08% female. Income was imbalanced: 75.92% earned `<=50K` and 24.08% earned `>50K`. Every education category mapped consistently to one education-num value.

**Cleaning decisions:** Missing categorical values were imputed as `Unknown` to preserve observations without inventing a likely category. No numeric imputation was necessary; median imputation would be used if numeric missingness appeared. Categorical whitespace was stripped. Exact duplicates were reported but retained for auditability, and label-conflicting records were not treated as automatic duplicates.